atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/internalschema/TransactionSchemaManager.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * (c) Copyright 2018 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.internalschema;
+
+import java.util.Optional;
+
+import com.palantir.atlasdb.coordination.CoordinationService;
+import com.palantir.atlasdb.coordination.ValueAndBound;
+import com.palantir.atlasdb.keyvalue.impl.CheckAndSetResult;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+
+public class TransactionSchemaManager {
+    private final CoordinationService<InternalSchemaMetadata> coordinationService;
+
+    public TransactionSchemaManager(CoordinationService<InternalSchemaMetadata> coordinationService) {
+        this.coordinationService = coordinationService;
+    }
+
+    public int getTransactionsSchemaVersion(long timestamp) {
+        Optional<Integer> possibleVersion =
+                extractTimestampVersion(coordinationService.getValueForTimestamp(timestamp), timestamp);
+        while (!possibleVersion.isPresent()) {
+            CheckAndSetResult<ValueAndBound<InternalSchemaMetadata>> casResult = tryPerpetuateExistingState();
+            possibleVersion = extractTimestampVersion(casResult.existingValues()
+                    .stream()
+                    .filter(valueAndBound -> valueAndBound.bound() >= timestamp)
+                    .findAny(),
+                    timestamp);
+        }
+        return possibleVersion.get();
+    }
+
+    public void installNewTransactionsSchemaVersion(int newVersion) {
+        coordinationService.tryTransformCurrentValue(valueAndBound -> {
+            if (!valueAndBound.value().isPresent()) {
+                throw new SafeIllegalStateException("Persisted value is empty, which is unexpected.");
+            }
+
+            InternalSchemaMetadata internalSchemaMetadata = valueAndBound.value().get();
+            return InternalSchemaMetadata.builder()
+                    .from(internalSchemaMetadata)
+                    .timestampToTransactionsTableSchemaVersion(
+                            installNewVersionInMap(
+                                    internalSchemaMetadata.timestampToTransactionsTableSchemaVersion(),
+                                    valueAndBound.bound() + 1,
```

Comment:
Actually in this specific case I don't think it needs to be coordinated; what this does is that it sets the old transactions to stop at the old validity bound and, implicitly, the new one to stop at the next bound.

e.g. we previously had `[1, +inf):1` with validity up to timestamp `500`; post-transform this becomes `[1, 501):1,[501,+inf):2` with validity up to something, but I don't quite care whether that's `700`, `1000` or `314159265`. Might be better to discuss this offline - not sure I follow

Reply:
That said, I did think about whether in general it might be useful for transforms to be able to take the next bound as an argument as well - just that for this specific one I don't think it's needed, unless I'm missing something

Reply:
Looks good, I had confused myself. We use `valueAndBound.bound() + 1` as the new cutover, which I had misread as potentially being some timestamp that could be less than the current bound.